Updated moonshine modelcard (#38711)
huggingface/transformers: docs/source/en/model_doc/moonshine.md

-->

<div style="float: right;">
    <div class="flex flex-wrap space-x-1">
          <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-DE3412?style=flat&logo=pytorch&logoColor=white">
          <img alt="FlashAttention" src="https://img.shields.io/badge/%E2%9A%A1%EF%B8%8E%20FlashAttention-eae0c8?style=flat">
          <img alt="SDPA" src="https://img.shields.io/badge/SDPA-DE3412?style=flat&logo=pytorch&logoColor=white">
    </div>
</div>

# Moonshine

[Moonshine](https://huggingface.co/papers/2410.15608) is a speech recognition model that is optimized for real-time transcription and voice command. Instead of using traditional absolute position embeddings, Moonshine uses Rotary Position Embedding (RoPE).

Moonshine is trained on speech segments of various lengths, but without using zero-padding, leading to greater efficiency for the encoder during inference time.

> You can find all the Moonshine checkpoints on the [Hub](https://huggingface.co/models?search=moonshine).

Reviewer: ```suggestion
You can find all the original Moonshine checkpoints under the [Useful Sensors](https://huggingface.co/UsefulSensors) organization.
```

> [!TIP]
> Click on the Moonshine models in the right sidebar for more examples of how to apply Moonshine to different speech recognition tasks.

The example below demonstrates how to generate a transcription based on an audio file with [`Pipeline`] or the [`AutoModel`] class.

<hfoptions id="usage">
<hfoption id="Pipeline">

```py
# uncomment to install ffmpeg which is needed to decode the audio file
# !brew install ffmpeg

from transformers import pipeline

asr = pipeline("automatic-speech-recognition", model="UsefulSensors/moonshine-base")

result = asr("path_to_audio_file")

#Prints the transcription from the audio file
print(result["text"])
```

</hfoption>
<hfoption id="AutoModel">

```py

> # uncomment to install librosa which is used for audio and music anlaysis. It is used to preprocess the data.

Reviewer: ```py
# pip install datasets
import torch
from datasets import load_dataset
from transformers import AutoProcessor, MoonshineForConditionalGeneration

processor = AutoProcessor.from_pretrained(
    "UsefulSensors/moonshine-base",
)
model = MoonshineForConditionalGeneration.from_pretrained(
    "UsefulSensors/moonshine-base",
    torch_dtype=torch.float16,
    device_map="auto",
    attn_implementation="sdpa"
).to("cuda")

ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", split="validation")
audio_sample = ds[0]["audio"]

input_features = processor(
    audio_sample["array"],
    sampling_rate=audio_sample["sampling_rate"],
    return_tensors="pt"
)
input_features = input_features.to("cuda", dtype=torch.float16)

predicted_ids = model.generate(**input_features, cache_implementation="static")
transcription = processor.batch_decode(predicted_ids, skip_special_tokens=True)
transcription[0]
```

# !pip install librosa
import torch
from transformers import AutoProcessor, AutoModelForSpeechSeq2Seq

processor = AutoProcessor.from_pretrained("UsefulSensors/moonshine-tiny")
model = AutoModelForSpeechSeq2Seq.from_pretrained("UsefulSensors/moonshine-tiny")

audio_array, sr = librosa.load("pathToFile", sr=16000)
inputs = processor(audio_array, return_tensors="pt", sampling_rate=16000)

generated_ids = model.generate(**inputs, max_new_tokens=256)
transcription = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]

print(f"Transcription: '{transcription}'")
```
</hfoption>
</hfoptions>

> ## Notes

Reviewer: ```suggestion
```

- Moonshine improves upon Whisper's architecture:
  1. It uses SwiGLU activation instead of GELU in the decoder layers
  2. Most importantly, it replaces absolute position embeddings with Rotary Position Embeddings (RoPE). This allows Moonshine to handle audio inputs of any length, unlike Whisper which is restricted to fixed 30-second windows.

> - A guide for automatic speech recognition can be found [here](../tasks/asr)

Reviewer: ```suggestion
```
